phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexUpgradeTool.java

```diff
@@ -0,0 +1,459 @@
+                FileHandler fh = new FileHandler(logFile);
+                fh.setFormatter(new SimpleFormatter());
+                LOGGER.addHandler(fh);
+            }
+            if (dryRun) {
+                LOGGER.info("This is the beginning of the tool with dry run.");
+            }
+            LOGGER.info("Index Upgrade tool initiated: "+ String.join(",", args));
+
+            prop.put(Indexer.INDEX_BUILDER_CONF_KEY, PhoenixIndexBuilder.class.getName());
+            prop.put(NonTxIndexBuilder.CODEC_CLASS_NAME_KEY, PhoenixIndexCodec.class.getName());
+
+            connectionURL = cmdLine.getOptionValue(CONNECTION_STRING_OPTION.getOpt());
+
+            if (connectionURL == null) {
+                connectionURL = "jdbc:phoenix:"+host+":"+port;
+            }
+
+            if (tables == null) {
+                String tablesCsv = cmdLine.getOptionValue(TABLE_CSV_FILE_OPTION.getOpt());
+                tables = new String(Files.readAllBytes(Paths.get(tablesCsv)));
+            }
+            LOGGER.info("list of tables/indexes passed: "+ tables);
+
+            if (tables == null) {
+                LOGGER.severe("tables' list is not available; use -tb or -f option");
+                System.exit(1);
+            }
+
+            HashMap<String, ArrayList<String>> tablesAndIndexes = getTablesAndIndexes(tables);
+
+            if (operation.equalsIgnoreCase(UPGRADE_OP)) {
+                LOGGER.info("Executing upgrade");
+                upgradeConsistencyFeature(tablesAndIndexes);
+            } else if (operation.equalsIgnoreCase(ROLLBACK_OP)) {
+                LOGGER.info("Executing rollback");
+                rollbackConsistencyFeature(tablesAndIndexes);
+            } else {
+                throw new IllegalStateException("Invalid option provided for "+OPERATION_OPTION.getOpt()+" expected values: {upgrade, rollback}");
+            }
+
+        } catch (IOException e) {
+            LOGGER.severe("Something went wrong "+e);
+        }
+    }
+
+    private static void upgradeConsistencyFeature(HashMap<String, ArrayList<String>> tablesAndIndexes) {
+
+        try {
+            Connection conn = DriverManager.getConnection(connectionURL, new Properties());
+            ConnectionQueryServices queryServices = conn.unwrap(PhoenixConnection.class).getQueryServices();
+            Configuration conf = queryServices.getConfiguration();
+            tablesAndIndexes.entrySet().parallelStream().forEach(entry -> {
+
+                try (Admin admin = queryServices.getAdmin()) {
+
+                    String fullTableName = entry.getKey();
+                    PTable dataTable = PhoenixRuntime.getTableNoCache(conn, fullTableName);
+                    LOGGER.fine("Executing upgrade for "+ fullTableName);
+
+                    boolean mutable = !(dataTable.isImmutableRows());
+
+                    checkAndWaitForMajorCompaction(admin, entry);
+
+                    resetMajorCompaction(admin, entry);
```

Comment:
Don't we need to disable major compaction first and then wait for currently running ones to end? Otherwise, there's a slim window that another one can start between the two.

Reply:
@priyankporwal - what's tricky is that we're running on the mnds node and we don't necessarily know if a major compaction's going on in a region, I don't think.

Reply:
Hmmm ... then I'm not even sure why we should rely on getCompactionState() if it does not give us accurate status of ongoing major compactions.I see there is getCompactionStateForRegion() - but that would require us to enumerate all the regions - and that may itself pose race when region splits and merges are happening.  Is there any better way?

Reply:
As discussed offline, PHOENIX-5373 will consider the scenario of stale rows and upgrade can avoid hitting race condition if any due to compaction.